Use std::filesystem instead of rcpputils::fs

r7vme opened the pull request:
Based on PRs https://github.com/ros2/rosbag2/pull/1478 and https://github.com/ros2/rosbag2/pull/1479

Use std::filesystem instead of rcpputils::fs (https://github.com/ros2/rcpputils/issues/164) in all packages of rosbag2.

~~The only place where it is still used is TemporaryDirectoryFixture in https://github.com/ros2/rosbag2/blob/1564687de2d10d5bce32dcb9b38ec5c9239416a4/rosbag2_test_common/include/rosbag2_test_common/temporary_directory_fixture.hpp#L34~~

~~Because there is no alternative in the standard library~~

yoneken committed:
Remove rcpputils::fs dependency from rosbag2_cpp

Signed-off-by: Kenta Yonekura <yoneken@ieee.org>

yoneken committed:
Remove rcpputils::fs dependency from tests

Signed-off-by: Kenta Yonekura <yoneken@ieee.org>

r7vme committed:
Replace rcpputils::fs with std::filesystem

Signed-off-by: Roman Sokolkov <rsokolkov@gmail.com>

clalancette commented:
> Because there is no alternative in the standard library

For what it is worth, the only users of `rcpputils::fs::create_temp_directory` are here in rosbag2: https://github.com/ros2/rosbag2/blob/1564687de2d10d5bce32dcb9b38ec5c9239416a4/rosbag2_test_common/include/rosbag2_test_common/temporary_directory_fixture.hpp#L34 and https://github.com/ros2/rosbag2/blob/1564687de2d10d5bce32dcb9b38ec5c9239416a4/rosbag2_transport/test/rosbag2_transport/test_rewrite.cpp#L54 .

Since the goal is to eventually remove `rcpputils::fs`, I think it is reasonable enough to copy https://github.com/ros2/rcpputils/blob/292be119c13edb8baa1e78c348637c460ad7a846/src/filesystem_helper.cpp#L310-L342 somewhere into this repository.

r7vme commented:
@clalancette thanks, done. Do we run CI on Windows? Just want to be sure it works there too

@MichaelOrlov could you also please take a look?

fujitatomoya commented on rosbag2_cpp/src/rosbag2_cpp/reindexer.cpp:
```diff
@@ -99,33 +97,34 @@ bool Reindexer::compare_relative_file(
-  if (dir_iter == nullptr) {
+  if (std::filesystem::is_empty(base_folder)) {
     throw std::runtime_error("Empty directory.");
   }
```
we should check this in the 1st place?

r7vme replied:
done, moved to the beginning of the function

fujitatomoya commented on rosbag2_cpp/src/rosbag2_cpp/reindexer.cpp:
```diff
@@ -99,33 +97,34 @@ bool Reindexer::compare_relative_file(
-    ROSBAG2_CPP_LOG_DEBUG_STREAM("Found file: " << found_file.string());
+  for (; first != last; ++first) {
+    auto found_file = first->path().filename();
+    ROSBAG2_CPP_LOG_DEBUG_STREAM("Found file: " << found_file.generic_string());
```
how about the following? so that we can delete the above local variables in stack?

```suggestion
  for (const auto& entry; std::filesystem::directory_iterator(base_folder)) {
    auto found_file = entry.path().filename();
    ROSBAG2_CPP_LOG_DEBUG_STREAM("Found file: " << found_file.generic_string());
```

r7vme replied:
done, thanks

fujitatomoya commented on rosbag2_test_common/include/rosbag2_test_common/temporary_directory_fixture.hpp:
```diff
@@ -17,26 +17,61 @@
+{
+  const auto template_path = base_name + "XXXXXX";
+  std::string full_template_str = (parent_path / template_path).generic_string();
+  std::filesystem::create_directories(parent_path);
```
std::filesystem::create_directories could fail and return false.

```suggestion
  if(!std::filesystem::create_directories(parent_path)) {
    std::error_code ec{errno, std::system_category()};
    errno = 0;
    throw std::system_error(ec, "could not create the parent directory");
  }
```

r7vme replied:
It works same as before no need to handle `false` return code.

`std::filesystem::create_directories(parent_path)` will:
1. return true if has actually created directories
2. return false if directory already exists
3. throw exception in other cases (e.g. permisssion denied or not a directory)

This behaviour aligned with the version from rcpputils. Except may be that std::filesystem_error and not std::system_error is thrown. This function only used in tests, so I think we can skip adding tests.

fujitatomoya replied:
i see! thanks for the info!

r7vme commented:
@fujitatomoya thanks for the comments, addressed

MichaelOrlov commented on rosbag2_compression/src/rosbag2_compression/sequential_compression_writer.cpp:
```diff
@@ -14,6 +14,7 @@
 
 #include "rosbag2_compression/sequential_compression_writer.hpp"
 
+#include <filesystem>
```
Please add new includes in the alphabetical order.

r7vme replied:
done

MichaelOrlov commented on rosbag2_compression/src/rosbag2_compression/sequential_compression_writer.cpp:
```diff
@@ -24,7 +25,6 @@
 #include "rcpputils/asserts.hpp"
-#include "rcpputils/filesystem_helper.hpp"
 
 #include "rcutils/filesystem.h"
```
It seems `#include "rcpputils/filesystem.h"` also became unused and could be deleted.

r7vme replied:
done

MichaelOrlov commented on rosbag2_compression/src/rosbag2_compression/sequential_compression_writer.cpp:
```diff
@@ -292,14 +292,14 @@ void SequentialCompressionWriter::compress_file(
   ROSBAG2_COMPRESSION_LOG_INFO_STREAM("Compressing file: " << file_relative_to_pwd.string());
 
-  if (file_relative_to_pwd.exists() && file_relative_to_pwd.file_size() > 0u) {
+  if (std::filesystem::exists(file_relative_to_pwd) &&
+    std::filesystem::file_size(file_relative_to_pwd) > 0u)
+  {
```
Neatpick. (Optional)
I would add 
```
namespace fs = std::filesystem;
```
at the beginning of the method to shorten expressions related to the filesystem calls.

r7vme replied:
I would avoid that if possible, [google style guide](https://google.github.io/styleguide/cppguide.html#Namespaces) also mentions that

```
Do not use using-directives (e.g., using namespace foo)
```

MichaelOrlov replied:
> I would avoid that if possible, [google style guide](https://google.github.io/styleguide/cppguide.html#Namespaces) also mentions that
> Do not use using-directives (e.g., using namespace foo)

My proposal is different from the style guide. There is a big difference in the 
```
using namespace std::filesystem;
```
and proposed
```
namespace fs = std::filesystem;
```

r7vme replied:
got it, done

MichaelOrlov commented on rosbag2_compression/src/rosbag2_compression/sequential_compression_writer.cpp:
```diff
@@ -318,7 +318,7 @@ void SequentialCompressionWriter::compress_file(
+    if (!std::filesystem::remove(file_relative_to_pwd)) {
       ROSBAG2_COMPRESSION_LOG_ERROR_STREAM(
         "Failed to remove original pre-compressed bag file: \"" <<
           file_relative_to_pwd.string() << "\". This should never happen - but execution " <<
```
Need to use a non-throwing overload of the `fs::remove(..)` here. 
```suggestion
    if (std::error_code ec; !fs::remove(file_relative_to_pwd, ec)) {
      ROSBAG2_COMPRESSION_LOG_ERROR_STREAM(
        "Failed to remove original pre-compressed bag file: \"" <<
          file_relative_to_pwd.string() << "\". " << ec.message() <<
          ". This should never happen - but execution " <<
```

r7vme replied:
done

MichaelOrlov commented on rosbag2_compression_zstd/src/rosbag2_compression_zstd/zstd_decompressor.cpp:
```diff
@@ -12,15 +12,14 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#include <filesystem>
```
Please add new includes in the alphabetical order.

r7vme replied:
done

MichaelOrlov commented on rosbag2_compression_zstd/src/rosbag2_compression_zstd/zstd_decompressor.cpp:
```diff
@@ -45,8 +44,8 @@ ZstdDecompressor::~ZstdDecompressor()
-  const auto uri_path = rcpputils::fs::path{uri};
-  const auto decompressed_uri = rcpputils::fs::remove_extension(uri_path).string();
+  auto uri_path = std::filesystem::path{uri};
+  const auto decompressed_uri = uri_path.replace_extension().generic_string();
```
```suggestion
  auto decompressed_uri = uri_path.replace_extension().generic_string();
```
CLion showing a warning when we are trying to return `decompressed_uri` from method
> Clang-Tidy: Constness of 'decompressed_uri' prevents automatic move 

i.e. const prevents copy-elision for the returning value

r7vme replied:
nice, fixed

MichaelOrlov commented on rosbag2_compression_zstd/test/rosbag2_compression_zstd/test_zstd_compressor.cpp:
```diff
@@ -12,6 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#include <filesystem>
```
Please add new includes in the alphabetical order.

r7vme replied:
done

MichaelOrlov commented on rosbag2_compression_zstd/package.xml:
```diff
@@ -15,7 +15,6 @@
 
   <depend>pluginlib</depend>
-  <depend>rcpputils</depend>
   <depend>rcutils</depend>
```
Thanks for the deleting unused dependencies.
We can delete dependencies from the `rcutils` as well.

r7vme replied:
it still uses `rcutils`, e.g. in https://github.com/ros2/rosbag2/blob/a2283ab47eaa6d931af9e1b3117e0689acd5a40e/rosbag2_compression_zstd/src/rosbag2_compression_zstd/compression_utils.cpp#L98

clalancette replied:
Not directly related to this PR, but `compression_utils.cpp` should `#include <rcutils/rcutils_ret.h>`.

r7vme replied:
@clalancette good catch, added

MichaelOrlov commented on rosbag2_cpp/test/rosbag2_cpp/test_sequential_reader.cpp:
```diff
@@ -66,12 +65,13 @@ class SequentialReaderTest : public Test
     bag_file_2_path_ = relative_file_path_ / "bag_file2";
-    metadata_.relative_file_paths = {bag_file_1_path_.string(), bag_file_2_path_.string()};
+    metadata_.relative_file_paths =
+    {bag_file_1_path_.generic_string(), bag_file_2_path_.generic_string()};
```
Neatpick:
```suggestion
    metadata_.relative_file_paths = {
      bag_file_1_path_.generic_string(),
      bag_file_2_path_.generic_string()
    };
```

r7vme replied:
done

MichaelOrlov commented on rosbag2_cpp/src/rosbag2_cpp/readers/sequential_reader.cpp:
```diff
@@ -307,7 +307,7 @@ std::string SequentialReader::get_current_file() const
   auto current_file = get_current_file();
-  auto current_uri = rcpputils::fs::remove_extension(current_file);
+  auto current_uri = std::filesystem::path(current_file).stem();
   return current_uri.string();
```
Need return `current_uri.generic_string()`

r7vme replied:
done

MichaelOrlov commented on rosbag2_transport/test/rosbag2_transport/test_play_seek.cpp:
```diff
@@ -14,6 +14,7 @@
 
 #include <gmock/gmock.h>
 
+#include <filesystem>
```
Please add new includes in the alphabetical order.

r7vme replied:
done

MichaelOrlov commented on rosbag2_transport/test/rosbag2_transport/test_rewrite.cpp:
```diff
@@ -75,11 +76,10 @@ class TestRewrite : public Test, public WithParamInterface<std::string>
 
   ~TestRewrite()
   {
-    // rcpputils::fs::remove_all(output_dir_);
```
I think we need `fs::remove_all(output_dir)` here, since we creating a temp dir in the constructor.  
It seems It was commented out by mistake.

r7vme replied:
done

MichaelOrlov reviewed (changes requested):
@r7vme Thank you for picking up and finishing this PR. It turns out to be a large PR.
In general, it looks good to me. However, I have a few findings which need to be addressed.
Among a few minor findings which I pointed out in the inline comments there is some missed usage of the `rcutils_calculate_directory_size(uri.c_str(), &metadata.bag_size, allocator))` in the 
https://github.com/ros2/rosbag2/blob/a2283ab47eaa6d931af9e1b3117e0689acd5a40e/rosbag2_storage/src/rosbag2_storage/metadata_io.cpp#L38-L48
which could be replaced by using https://en.cppreference.com/w/cpp/filesystem/file_size instead

r7vme commented:
@MichaelOrlov addressed your comments here https://github.com/ros2/rosbag2/pull/1576/commits/0b53db4d7595303d3164878fcc36addb3e3b1708

Should I squash last 4 commits into one?

clalancette commented on rosbag2_test_common/include/rosbag2_test_common/temporary_directory_fixture.hpp:
```diff
@@ -17,26 +17,61 @@
+/**
+ * @brief std::filesystem version of create_temp_directory from rcpputils/src/filesystem_helper.cpp
+ */
+std::filesystem::path create_temp_directory(
+  const std::string & base_name,
+  const std::filesystem::path & parent_path = std::filesystem::temp_directory_path())
+{
+  const auto template_path = base_name + "XXXXXX";
+  std::string full_template_str = (parent_path / template_path).generic_string();
+  std::filesystem::create_directories(parent_path);
+
+#ifdef _WIN32
+  errno_t errcode = _mktemp_s(&full_template_str[0], full_template_str.size() + 1);
+  if (errcode) {
+    std::error_code ec(static_cast<int>(errcode), std::system_category());
+    throw std::system_error(ec, "could not format the temp directory name template");
+  }
+  const std::filesystem::path final_path{full_template_str};
+  std::filesystem::create_directory(final_path);
+#else
+  const char * dir_name = mkdtemp(&full_template_str[0]);
+  if (dir_name == nullptr) {
+    std::error_code ec{errno, std::system_category()};
+    errno = 0;
+    throw std::system_error(ec, "could not format or create the temp directory");
+  }
+  const std::filesystem::path final_path{dir_name};
+#endif
+
+  return final_path;
+}
```
I'm very sorry about this, but I now think that we should remove this and keep it in `rcpputils`.  The main reason is that I found another use for this function in https://github.com/ros2/rcl_logging/pull/115 , so even when we remove most of https://github.com/ros2/rcpputils/blob/292be119c13edb8baa1e78c348637c460ad7a846/src/filesystem_helper.cpp , we will still want to keep this function as a utility not offered by the STL.

Again, apologies for my short-sightedness.

r7vme replied:
reverted

MichaelOrlov commented:
@r7vme Thank you for addressing most of the comments from the review. 
It seems one comment was missed
> Among a few minor findings which I pointed out in the inline comments there is some missed usage of the `rcutils_calculate_directory_size(uri.c_str(), &metadata.bag_size, allocator))` in the
> 
> https://github.com/ros2/rosbag2/blob/a2283ab47eaa6d931af9e1b3117e0689acd5a40e/rosbag2_storage/src/rosbag2_storage/metadata_io.cpp#L38-L48
> 
> 
> which could be replaced by using https://en.cppreference.com/w/cpp/filesystem/file_size instead

r7vme commented:
@MichaelOrlov regarding `rcutils_calculate_directory_size`. Unfortunately there is no alteranative in std::filesystem. std::filesystem::file_size will return  `filesystem error: cannot get file size: Is a directory` on Linux if run for directrory. I would leave this piece as is. But I definitely missed this comment in the first place.

r7vme commented:
@clalancette @MichaelOrlov looks like `Rpr__rosbag2__ubuntu_noble_amd64 ` [fails](https://build.ros2.org/job/Rpr__rosbag2__ubuntu_noble_amd64/6/consoleFull) on all latest PRs with `The command '/bin/sh -c pip3 install -U setuptools==59.6.0' returned a non-zero code: 1`

MichaelOrlov commented:
@r7vme Yeah, we are aware of it. This is a generic issue due to the transition to the Ubuntu 24.04 on Rolling. This issue is going to be addressed sometime soon.

MichaelOrlov commented:
@r7vme As regards
> Should I squash last 4 commits into one?

Please do not squash commits in the future. It is very helpful for the reviewer to see changes in the separate commits.
We have automatically mandatory squash for all commits in the PR during the final merge to the target branch.

MichaelOrlov reviewed (approved):
@r7vme Thanks, LGTM!

MichaelOrlov commented:
Gist: https://gist.githubusercontent.com/MichaelOrlov/f323f6647c62a2395c5006aa68777d14/raw/fa6e1966d9c3fc82d4d4eeb990dd9e3461fe97b8/ros2.repos
BUILD args:  --packages-above-and-dependencies rosbag2_compression rosbag2_compression_zstd rosbag2_cpp  rosbag2_storage rosbag2_test_common rosbag2_transport rosbag2_tests
TEST args:  --packages-above rosbag2_compression rosbag2_compression_zstd rosbag2_cpp  rosbag2_storage rosbag2_test_common rosbag2_transport rosbag2_tests
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/13467
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=20479)](http://ci.ros2.org/job/ci_linux/20479/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=14881)](http://ci.ros2.org/job/ci_linux-aarch64/14881/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=21225)](http://ci.ros2.org/job/ci_windows/21225/)

r7vme commented:
@MichaelOrlov thanks, is there anything required from my side? Or you can restart failed CI job Rpr__rosbag2__ubuntu_noble_amd64 ?

clalancette commented:
> @MichaelOrlov thanks, is there anything required from my side? Or you can restart failed CI job Rpr__rosbag2__ubuntu_noble_amd64 ?

These aren't going to work right at the moment.  We are close to making them work, but need a bit more time.  The regular CI above is good enough.

MichaelOrlov commented:
@r7vme Some tests failing on Windows CI https://ci.ros2.org/job/ci_windows/21225/  
At first glance, it seems failure relates to the filesystem operation.
```
C:\ci\ws\src\r7vme\rosbag2\rosbag2_cpp\test\rosbag2_cpp\test_sequential_reader.cpp:224
Expected equality of these values:
  closed_file
    Which is: "C:/Users/ContainerAdministrator/AppData/Local/Temp/some/folder/bag_file1"
  bag_file_1_path_.string()
    Which is: "C:/Users/ContainerAdministrator/AppData/Local/Temp/some/folder\\bag_file1"

C:\ci\ws\src\r7vme\rosbag2\rosbag2_cpp\test\rosbag2_cpp\test_sequential_reader.cpp:225
Expected equality of these values:
  opened_file
    Which is: "C:/Users/ContainerAdministrator/AppData/Local/Temp/some/folder/bag_file2"
  bag_file_2_path_.string()
    Which is: "C:/Users/ContainerAdministrator/AppData/Local/Temp/some/folder\\bag_file2"
```
Could you please take a look at failures and test use cases? It seems we have forgotten to use `.string_generic()` instead of `.string()` for `fs::path` somewhere.

r7vme committed:
Use .generic_string() in tests

Signed-off-by: Roman Sokolkov <rsokolkov@gmail.com>

MichaelOrlov commented:
Re-run CI afer attempt to fix test errors on Windows build
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/13484
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=20496)](http://ci.ros2.org/job/ci_linux/20496/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=14898)](http://ci.ros2.org/job/ci_linux-aarch64/14898/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=21245)](http://ci.ros2.org/job/ci_windows/21245/)

MichaelOrlov commented:
@r7vme We have one more test failure on Windows
```
[ RUN      ] StorageTestFixture.storage_preset_profile_applies_over_defaults
[INFO] [1710819321.702947900] [rosbag2_storage]: Opened database 'C:/Users/ContainerAdministrator/AppData/Local/Temp/tmp_test_dir_a05376/rosbag.db3' for READ_WRITE.
[       OK ] StorageTestFixture.storage_preset_profile_applies_over_defaults (32 ms)
[ RUN      ] StorageTestFixture.throws_on_invalid_pragma_in_config_file
```
Could you please take a look on it?

r7vme committed:
Revert remove_all from fixture

Signed-off-by: Roman Sokolkov <rsokolkov@gmail.com>

r7vme commented:
@MichaelOrlov I [reverted](https://github.com/ros2/rosbag2/pull/1576/commits/d689b99e913343e91a95ce182e500ea1f0b3f482) remove_all in test fixture. This part is used in failed test. Let's see it if helps. No other ideas unforunately

MichaelOrlov commented:
Trying to re-run Windows CI after reverting remove_all in the test fixture to see if it is a root cause of the CI failure
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=21248)](http://ci.ros2.org/job/ci_windows/21248/)

r7vme commented:
@clalancette could you please take another look? your comment was addressed

clalancette reviewed (approved):
> @MichaelOrlov I [reverted](https://github.com/ros2/rosbag2/pull/1576/commits/d689b99e913343e91a95ce182e500ea1f0b3f482) remove_all in test fixture. This part is used in failed test. Let's see it if helps. No other ideas unforunately

I have to say that I'm surprised that this fixed the problem.  We are using a similar pattern in https://github.com/ros2/rcl_logging/blob/rolling/rcl_logging_spdlog/test/test_logging_interface.cpp , and it is working fine by using `std::filesystem::remove_all`.

That said, I'm still approving since I'm OK with putting this in as-is.  It just means that we won't be able to remove `rcpputils::fs` until we do another PR to fix whatever bug `std::filesystem::remove_all` is hitting.

r7vme commented:
@MichaelOrlov I guess can be merged finally?

MichaelOrlov commented:
@r7vme Yes we can merge I think. However, please hold off for a couple of days. I need to wrap up some other things and I want to mitigate the potential merge or rebase conflicts.
